index.js

```diff
@@ -173,22 +188,28 @@ DB.prototype.del = function (key, opts, cb) {
   })
 }
 
-DB.prototype._del = function (key, opts, cb) {
+// OsmVersion, Opts -> [OsmBatchOp]
+DB.prototype._getDocumentDeletionBatchOps = function (key, opts, cb) {
   var self = this
-  self.kv.get(key, function (err, values) {
+  self.kv.get(key, function (err, docs) {
     if (err) return cb(err)
-    // Filter deletions & map to old expected format.
-    values = filterObj(values, function (key, value) {
-      return !value.deleted
-    })
-    values = mapObj(values, function (key, value) {
-      return value.value
+
+    docs = mapObj(docs, function (version, value) {
+      if (value.deleted) {
+        return {
+          id: key,
+          version: version,
+          deleted: true
+        }
+      } else {
+        return value.value
+      }
     })
 
     var fields = {}
-    var links = opts.keys || Object.keys(values)
+    var links = opts.keys || Object.keys(docs)
     links.forEach(function (ln) {
-      var v = values[ln] || {}
+      var v = docs[ln] || {}
```

Comment:
Changes via the OSM API always have opts.links set, since clients like iD editor always include the version of the element that is modified in the changeset. It is quite possible that the current head of the db changes between a client getting changes and submitting a changeset. This could happen in a future scenario where data is live replicated in the background, or where a user group might have osm-p2p-db on a local server with various clients all editing against the same server, or where a single machine might have multiple osm-p2p-db clients.

In this case, for a given node Av1, iD might make a bbox request, the user would edit Av1, then, before they submit changes, another user or replication might create Av1 <--- Av2. When iD submits a change opts.links would be [v1], but self.kv.get() would return docs = { v2: .... }.